src/App/Pages/Accounts/TwoFactorPageViewModel.cs

```diff
@@ -208,6 +224,51 @@ public void Load()
             ShowContinue = !(SelectedProviderType == null || DuoMethod || Fido2Method);
         }
 
+        private async Task DuoFramelessAuthenticateAsync()
+        {
+            await _deviceActionService.ShowLoadingAsync(AppResources.Validating);
+            var providerData = _authService.TwoFactorProvidersData[SelectedProviderType.Value];
+            var url = providerData["AuthUrl"] as string;
+            WebAuthenticatorResult authResult;
+            try
+            {
+                authResult = await WebAuthenticator.AuthenticateAsync(new WebAuthenticatorOptions
+                {
+                    Url = new Uri(url),
+                    CallbackUrl = new Uri("bitwarden://duo-callback")
+                });
+            }
+            catch (TaskCanceledException)
+            {
+                // user canceled
+                await _deviceActionService.HideLoadingAsync();
+                return;
+            }
+
+            string response = null;
+            if (authResult != null && authResult.Properties.TryGetValue("code", out var resultData))
+            {
+                response = Uri.UnescapeDataString(resultData);
+            }
+            if (!string.IsNullOrWhiteSpace(response))
+            {
+                Token = response;
+                await SubmitAsync(false);
+                return;
+            }
+
+            await _deviceActionService.HideLoadingAsync();
+            if (authResult != null && authResult.Properties.TryGetValue("error", out var resultError))
+            {
+                await _platformUtilsService.ShowDialogAsync(resultError, AppResources.AnErrorHasOccurred,
+                    AppResources.Ok);
+            }
+            else
+            {
+                await _platformUtilsService.ShowDialogAsync(AppResources.AnErrorHasOccurred, AppResources.Ok);
+            }
```

Comment:
🤔 IMO it's easier to follow when errors are first and we let the main flow continue on the main branch of the code, so I'd invert the code and move this part above inside `if (string.IsNullOrWhiteSpace(response))` (inverting the condition) and put the next here.
```c#
Token = response;
await SubmitAsync(false);
```
❓ Additionally, is `resultError` human readable and localized?

Reply:
Agreed! Moved error handling up before the continue of the flow. Also added some error logging and handling. Let me know what you think.

Reply:
Since I can't be sure of the `resultError` readability, I logged the error message and shown the user a generic error only.